packages/Button/src/Button.tsx

```diff
@@ -267,61 +281,124 @@ const baseStyle = css`
+  href?: null;
+}
+
+type CustomElementButtonProps = SharedButtonProps & {
+  as: React.ElementType<any>;
+  [key: string]: any;
+};
+
+type ButtonProps =
+  | LinkButtonProps
+  | ButtonButtonProps
+  | CustomElementButtonProps;
+
+function hasCustomRoot(props: ButtonProps): props is CustomElementButtonProps {
+  return (props as any).as != null;
+}
+
+function hasLinkRoot(
+  props: LinkButtonProps | ButtonButtonProps,
+): props is LinkButtonProps {
+  return props.href != null;
+}
+
+const omitProps = [
+  'as',
+  'className',
+  'disabled',
+  'size',
+  'variant',
+  'children',
+] as const;
+
+export default class Button extends Component<ButtonProps> {
+  static displayName = 'Button';
 
   static defaultProps = {
     variant: 'default',
     size: 'normal',
     className: '',
     children: null,
     disabled: false,
-    type: 'button',
   };
 
   render() {
-    const {
-      className,
-      children,
+    const { className, children, disabled, variant, size } = this.props;
+
+    const commonProps = {
+      className: cx(
+        baseStyle,
+        buttonSizes[size],
+        buttonVariants[variant],
+        className,
+      ),
       disabled,
-      variant,
-      size,
-      as,
-      href,
-      ...rest
-    } = this.props;
+      'aria-disabled': disabled,
+    };
 
-    let Root = href ? 'a' : 'button';
+    if (hasCustomRoot(this.props)) {
+      const Root = this.props.as;
+      const rest = omit(this.props, omitProps);
 
-    if (as) {
-      Root = as;
+      return (
+        <Root {...rest} {...commonProps}>
+          {children}
+        </Root>
+      );
     }
 
+    if (hasLinkRoot(this.props)) {
+      const rest: HTMLElementProps<'a'> = omit(this.props, omitProps);
+      return (
+        <a {...rest} {...commonProps}>
+          {children}
+        </a>
+      );
+    }
+
+    const rest: HTMLElementProps<'button'> = omit(this.props, omitProps);
+
+    // NOTE(JeT): The button's `type` will be overridden if it is in the passed-in props
     return (
-      <Root
-        {...rest}
-        href={href}
-        className={cx(
-          baseStyle,
-          buttonSizes[size],
-          buttonVariants[variant],
-          className,
-        )}
-        disabled={disabled}
-        aria-disabled={disabled}
-      >
+      <button type="button" {...rest} {...commonProps}>
         {children}
-      </Root>
+      </button>
     );
   }
 }
+
+/*
+NOTE(JeT):
+When this is defined inside the actual component, React will try to infer prop types from it,
+merging them together with ButtonProps (see LibraryManagedAttributes and MergePropTypes in @types/react).
+Unfortunately, this merging uses keyof, which appears to drop the [key: string] index signature, meaning TS won't 
+allow us to pass unrecognized props down to custom components when we're using the `as` prop. This workaround avoids
+the attempt at merging, while still getting runtime type-checking for non-TS consumers of the library.
+*/
```

Comment:
I want to draw special attention to this comment, because it was a weird confusing situation that I'd like to make sure I'm explaining well. Of special note, the two `@types/react` definitions I refer to:
[LibraryManagedAttributes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/932108dd26567dd5cfbdcc7b80b2c37e39af36b9/types/react/index.d.ts#L2818) and [MergePropTypes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/932108dd26567dd5cfbdcc7b80b2c37e39af36b9/types/react/index.d.ts#L2772)

Reply:
Good comment!